### ISSUE TITLE
V3 plan: fine-tune MSI-Net on UEyes — handoff brief

### ISSUE THREAD

**Issue by khawkins98** [khawkins98/foveacast-training#1]
A self-contained brief for the first substantive body of work in this repo. Written so a fresh Claude Code session (or a fresh human contributor) can land running without re-reading the entire Foveacast history.

## TL;DR

Produce a UI-content-aware saliency model for [Foveacast](https://github.com/khawkins98/Foveacast) by fine-tuning MSI-Net on the UEyes dataset. Ship the result as a tagged GitHub Release with an `.onnx` artefact attached. A separate integration PR on Foveacast then drops that artefact into `docs/models/` and wires it up.

**First action for a new session:** read this issue end-to-end, then read Foveacast's [`LEARNINGS.md`](https://github.com/khawkins98/Foveacast/blob/main/LEARNINGS.md) (specifically the 2026-04-16 entries on the model survey, the UMSI++ correction, and the companion-repo split) for the trajectory that got us here. Then check in before starting Phase 0 — there are still a couple of open decisions noted below.

## Why this work exists

- **Foveacast V1** (`0.1.0`/`0.1.1`) shipped [MSI-Net](https://github.com/alexanderkroner/saliency) (MIT, Kroner et al. 2020) via TensorFlow.js. Works, proven browser path, five quality presets.
- **Foveacast V2 spike** ([PR #4](https://github.com/khawkins98/Foveacast/pull/4)) swapped to [UNISAL](https://github.com/rdroste/unisal) (Apache 2.0, Droste et al. 2020) via ONNX Runtime Web. Architecturally cleaner (smaller, in-repo artefact, no GCS dependency), but empirically worse on UI content.
- **The benchmark that changed the plan.** Four qualitative comparisons in [`docs/spikes/comparison/`](https://github.com/khawkins98/Foveacast/tree/spike/unisal-onnx-research/docs/spikes/comparison) on the V2 branch — UNDRR page, YouTube home, and two ACS pages with **real eye-tracking ground truth** from the [UEyes CHI 2023 study](https://doi.org/10.1145/3544548.3581096). Both MSI-Net and UNISAL miss CTAs like the "Begin" button; MSI-Net at least tracks the rough structure of real fixation hotspots, UNISAL collapses to a central blob. The real diagnosis: both models are SALICON-trained, both are natural-scene-biased, neither knows what a button is. Picking between them is moving deck chairs.
- **The real fix is training on UI content.** The UEyes dataset (1,980 UI screenshots with real participant eye-tracking) is on Zenodo under CC BY 4.0. Fine-tuning MSI-Net (MIT) on UEyes (CC BY 4.0) is the shortest path to a UI-aware saliency model under clean licences end-to-end, with no author-contact gates and no Mamba CUDA fights (the V3 SUM path the original PRD sketched).

The full long-form trajectory is in Foveacast's `LEARNINGS.md`. The short version of "why fine-tuning and not X" lives in [`docs/spikes/model-survey-2026-04.md`](https://github.com/khawkins98/Foveacast/blob/spike/unisal-onnx-research/docs/spikes/model-survey-2026-04.md) — a survey of every plausible post-2023 candidate and why this one won.

## Repositories

- **This repo:** https://github.com/khawkins98/foveacast-training — training pipeline, dataset loader, model architecture, ONNX export. Produces the shipped artefact. MIT-licensed.
- **Foveacast (consumer):** https://github.com/khawkins98/Foveacast — buildless static web app. Consumes a tagged release from this repo at integration time.
- **MSI-Net upstream:** https://github.com/alexanderkroner/saliency — Alexander Kroner's reference implementation, MIT-licensed. Source for the model architecture and pretrained weights to fine-tune.
- **UEyes dataset:** https://zenodo.org/records/8010312 — Zenodo deposit, CC BY 4.0. 12.9 GB zip, 1,980 UI screenshots + 62-participant eye-tracking.

## Conventions expected in this repo

Mirroring Foveacast's — see its [`CLAUDE.md`](https://github.com/khawkins98/Foveacast/blob/main/CLAUDE.md) if a CLAUDE.md hasn't been copied into this repo yet.

- **Ample comments.** Every non-obvious branch gets a `# why:` inline. Module headers explain what lives there and why. Overriding any "self-documenting code" defaults; this is a research codebase people will pick up after the original context has decayed.
- **Conventional Commits** (`feat:`, `fix:`, `docs:`, `chore:`, `refactor:`, `test:`) with scopes where useful. Commit bodies explain *why*, not *what*.
- **Co-author trailer** when Claude helps:
  ```
  Co-Authored-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>
  ```
- **LEARNINGS.md is part of the workflow.** Any decision that took more than an hour to figure out, any dead end worth documenting, any "oh, that's how that actually works" moment gets a dated prose entry. Not a changelog; not a spec; a running log.
- **Humanizer pass on prose.** README, LEARNINGS, CHANGELOG, PR descriptions, commit messages — all read by humans. Strip AI-register tells before committing (no "seamlessly," "delightfully," "robust solution," "leverage," "best-in-class," rule-of-three filler).
- **Attribution is non-negotiable.** Every model, dataset, or code borrowing gets named credit in `README.md` and the relevant module header. See the README for the full chain.

## Plan, phased

Each phase is a checkbox. Each phase should be a small PR or a tagged commit sequence on a feature branch. Do not plow through all phases on a single mega-PR — each one has a real gate.

- [ ] **Phase 0 — Fetch and verify UEyes.** Download the 12.9 GB zip from Zenodo (manually once, or via a new `data/fetch.sh` helper). Unzip and produce a file-listing of what the deposit actually contains. Update `data/README.md` with the real directory layout (the current version is educated guessing). Confirm the train/val/test split the authors provide — or note that we need to create one. **Gate:** we should know the exact shape of inputs and ground truths before writing a loader for them.

- [ ] **Phase 1 — Decide the MSI-Net substrate.** Two options: (a) vendor Kroner's Keras/TF code verbatim and fine-tune in that stack; (b) port the architecture to PyTorch and load the pretrained weights across the framework gap. Preference is (b) — PyTorch has first-class Apple Silicon support via MPS, a cleaner ONNX export path, and keeps us on the same stack as Foveacast's V2 scripts. But it is meaningfully more work. **Gate:** user decision before sinking time.

- [ ] **Phase 2 — MSI-Net architecture in code.** Whichever path Phase 1 picks, land the model class with a verified forward pass. A unit test: run a fixed input through the model and the Kroner reference, diff the output. Should be pixel-close or within float tolerance.

- [ ] **Phase 3 — UEyes dataset loader.** PyTorch `Dataset` wrapping whatever shape Phase 0 found. Produce `(image_tensor, saliency_ground_truth)` pairs in the format MSI-Net's training loop expects (BGR, 0–255, specific input dimensions). Minimal data augmentation — this is a fine-tune on 1,980 images, not a from-scratch train on 10k.

- [ ] **Phase 4 — Prototype training run.** 100 images, 2 epochs, MPS backend (or CUDA if on a GPU box). Goal is "does the loss curve look plausible," not "is the model good." If this runs to completion in under 30 minutes and the loss decreases monotonically, Phase 5 is safe to start. If it NaN's, diverges, or doesn't fit in memory, that is the signal to stop and investigate.

- [ ] **Phase 5 — Full fine-tune.** Hyperparameters to start from: take Kroner's paper defaults, reduce learning rate by 10× for fine-tuning, 20–50 epochs with early stopping on validation metric. TensorBoard logging. Save best checkpoint on validation CC (correlation coefficient — standard saliency metric). Budget: 4–12 hours on an M4 MacBook Air, faster elsewhere.

- [ ] **Phase 6 — Evaluate quantitatively.** Saliency metrics (CC, KLD, NSS) on the UEyes held-out split. Compare against (a) stock MSI-Net without fine-tuning — does fine-tuning actually help? (b) published UMSI++ / commercial tool numbers if available, for context. If fine-tuning makes things worse, that is the call to stop and figure out what broke, not to ship.

- [ ] **Phase 7 — Evaluate qualitatively against Foveacast's benchmark.** Copy the four reference screenshots from Foveacast's [`docs/spikes/comparison/`](https://github.com/khawkins98/Foveacast/tree/spike/unisal-onnx-research/docs/spikes/comparison) into this repo's `benchmark/screenshots/`. Render the fine-tuned model's saliency map on each one. Compare to the ground-truth eye-tracking heatmaps for the two ACS pages. Does the new model actually hit the CTA and the sidebar the way real users do?

- [ ] **Phase 8 — ONNX export.** Pattern comes from Foveacast's [`scripts/unisal-onnx-export.py`](https://github.com/khawkins98/Foveacast/blob/spike/unisal-onnx-research/scripts/unisal-onnx-export.py). Same approach: load the best PyTorch checkpoint, wrap it in a 4-D-input adapter, `torch.onnx.export`, inline external data via `onnx.save_model(save_as_external_data=False)`, validate parity between PyTorch and `onnxruntime` CPU. Target artefact size: similar order to UNISAL's 12.5 MB.

- [ ] **Phase 9 — Release.** Tag a GitHub Release (e.g. `v0.1.0`). Attach the `.onnx` artefact to the release. Release notes should cover training config, dataset, evaluation numbers, and licence chain. Update this repo's `LEARNINGS.md` with the ship-day entry.

- [ ] **Phase 10 — Foveacast integration PR.** Cross-repo work. Open a PR against [Foveacast](https://github.com/khawkins98/Foveacast) that:
  - Drops the released `.onnx` into `docs/models/foveacast-v3/model.onnx`.
  - Points `docs/src/model/loader.js` at the new path (constants change; the structure carries over from the V2 spike branch that is already on the Foveacast repo).
  - Updates attribution footer and CHANGELOG.
  - Reruns Foveacast's Playwright tests against the new model.
  - Re-runs the four-screenshot comparison and adds the V3 outputs alongside the V1 / V2 versions that are already in `docs/spikes/comparison/`.

## Open decisions

Worth resolving with the maintainer before Phase 1:

1. **Port MSI-Net to PyTorch or vendor Keras/TF?** (Preference: port. But confirm.)
2. **Which MSI-Net input size to fine-tune?** Kroner ships five presets (48×64 → 240×320). V1 defaulted to 120×160. Probably fine-tune at one size and expose that single size in the export, mirroring V2's single-resolution UNISAL approach.
3. **Train/val/test split.** Take whatever split UEyes provides; if none, 80/10/10 with a fixed random seed so runs are reproducible.
4. **Target Apple Silicon MPS, CUDA, or both?** Code should auto-detect device; the question is which gets primary testing. M4 MacBook Air is the stated primary dev machine.

## Out of scope (explicitly)

- **Training from scratch.** Too little data; MSI-Net's SALICON pretraining is load-bearing.
- **Other architectures.** SUM, UMSI++, SalTR, DeepGaze were surveyed and not selected; see the [model survey](https://github.com/khawkins98/Foveacast/blob/spike/unisal-onnx-research/docs/spikes/model-survey-2026-04.md). If Phase 5 or 6 shows fine-tuned MSI-Net cannot beat stock MSI-Net, reopen that conversation. Otherwise, stay on task.
- **A web UI or demo for this repo.** This is a training pipeline. Foveacast is the UI. Keep the separation.
- **Dataset augmentation with synthetic UI.** Tempting, interesting, out of scope for the first training run.
- **Multi-task or multi-dataset training.** UEyes only for the first run.
- **Shipping weights through HuggingFace Hub.** GitHub Releases is enough for now; revisit if discoverability becomes a real product concern.

## How to close this issue

Close when Phase 9 lands and the Foveacast integration PR (Phase 10) is open. Reference both here. File follow-up issues for anything that slipped out of scope during the work — do not leave new questions hanging in this issue after it closes.
